### PR TITLE
Adding clFFT to LiFFT.

### PR DESCRIPTION
**Pull request by tdd11235813**
**Adding clFFT to LiFFT**
The basic functionality of clFFT is added to LiFFT:
- single- and double-precision,
- inplace and outplace,
- real and complex transforms
- if device memory only, then async behavior, otherwise transfers are synchronized
- see https://github.com/clMathLibraries/clFFT for feature set of clFFT

**Context management**
To work with OpenCL contexts and queues some simple policies are provided.

_ContextLocal_
- per-instance OpenCL and clFFT context
- prefers GPU device

```
using Context = LiFFT::libraries::clFFT::policies::ContextLocal;
using FFT = LiFFT::libraries::clFFT::ClFFT<Context>;
// each instance of FFT uses own context
```

_ContextGlobal_
- singleton OpenCL and clFFT context

```
using Context = LiFFT::libraries::clFFT::policies::ContextGlobal;
using FFT = LiFFT::libraries::clFFT::ClFFT<Context>;
// each instance of FFT uses same context
```

_ContextWrapper_
- user-defined OpenCL context [and queue]
- clFFT setup is optional

```
LiFFT::libraries::clFFT::policies::ContextLocal cl;
bool create_clfft_context=true;
LiFFT::libraries::clFFT::policies::ContextWrapper::wrap(
        cl.context(),
        cl.device(),
        [cl.queue(),]
        create_clfft_context);
using Context = LiFFT::libraries::clFFT::policies::ContextWrapper;
using FFT = LiFFT::libraries::clFFT::ClFFT<Context>;
```

**Tests**
_Environment_
- cuda/7.5.18
- opencl/1.2-4.4.0.117
- gcc/5.3.0
- boost/1.60.0
- clFFT/2.12.0

_State_
- [x] `ctest` on GPU (K20x, K80) and on CPUs.
- [x] Inplace policies (InplaceExplicit and InplaceForHost)
- [x] Context policies
- [ ] async calls & wrapped device pointer inplace/outplace (test case missing)

### CODE REVIEW

**Comment by tdd11235813** [comment]
... so, there are probably things you would like to change, just tell me what you think or what you would like to improve.
For the _wrapped device pointer_ test case I would like to introduce a testOpenCL.cpp or something like that, so I can do lib-specific tests.

File: CMakeLists.txt [hunk 1]
```diff
@@ -61,11 +75,17 @@ list(APPEND CUDA_NVCC_FLAGS_DEBUG "-O0 -g")
 if(${CMAKE_VERSION} VERSION_LESS 2.8.10)
     list(APPEND CUDA_NVCC_FLAGS "-ccbin ${CMAKE_C_COMPILER}")
 endif()
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
```
Comment (Flamefire) [inline CMakeLists.txt]: Can you add a comment, why this is required? Looks like a work around/hack for a specific version. If possible include link to bug report.

File: include/libLiFFT/libraries/clFFT/clFFT.hpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+     * Specifies that inplace transforms are only used when explicitly requested from the FFT
+     */
+    struct InplaceExplicit: std::false_type{};
+    /**
+     * Specifies that inplace transforms should be used if both memories reside on host
+     */
+    struct InplaceForHost: std::true_type{};
+
+    /**
+     * Wrapper for the OpenCL-Library that executes the FFT on GPU(s)
```
Comment (Flamefire) [inline include/libLiFFT/libraries/clFFT/clFFT.hpp]: Actually it executes the FFT on an `OpenCL device`, not only on GPUs :)

**Comment by Flamefire** [comment]
There are a couple if instances where you mention GPUs only. Please change them to "device" or "OpenCL device" as your implementations can run on more than only GPUs.
Also you write sometimes `clFFT` and sometimes `clfft`. Is it possible to have that consistent?
Writing library specific tests is a good idea, so `testOpenCL.cpp` is fine.

Besides that this looks great to me :+1: 

One suggestion: If we allow an additional, optional param to createPlan that is library-specific, we could pass an instance of ContextLocal to the planner. The main interface does not need to know its type at all and just passes it down to the library which would result in a compile-time error, if that type does not match the expected type (or maybe if there is none expected) This can be seen as an extension of the `ContextWrapper`, so you can submit plans to multiple queues. Or is this already possible (for more than 3 queues which can be "simulated" by using `ContextLocal`, `ContextGlobal` and `ContextWrapper`)

**Comment by tdd11235813** [comment]
thanks a lot for your positive answer and improvements, I am going to work on it and commit the changes to this pull request as soon as possible :)

**Comment by bussmann** [comment]
Nice work!

**Comment by tdd11235813** [comment]
Due to other tasks this project went a little bit under the wheels :P I am sorry! I started with some code refactoring recently. So for the "reboot" there will be a new PR, so I close this one now.

**Pull request closed without merging**